File: YeChanSong_코딩테스트_연습/힙/이중우선순위큐.py
import heapq

def solution(ops):
    answer = []
    minHq = list()
    maxHq = list()
    lenHq = 0
    for i in ops:
        inst, data = i.split()
        if lenHq == 0 and inst == 'D':
            continue
        v = int(data)
        if inst == 'I':
            heapq.heappush(minHq,v)
            heapq.heappush(maxHq,-v)
            lenHq += 1
            continue
        if inst == 'D' and v==1:
            itm = heapq.heappop(maxHq)
            minHq.remove(-itm)
            lenHq -=1
            continue
        else:
            itm = heapq.heappop(minHq)
            maxHq.remove(-itm)
            lenHq -=1
    if lenHq == 0:
        answer = [0,0]
    else:
        answer.append(-heapq.heappop(maxHq))
        answer.append(heapq.heappop(minHq))
    
    return answer

File: YeChanSong_코딩테스트_연습/힙/이중우선순위큐.java
import java.util.*;
class Solution {
    public int[] solution(String[] ops) {
        int[] answer;
        String[] tmp;
        int i;
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Comparator.reverseOrder());
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for(String str: ops){
            tmp = str.split(" ");
            if(tmp[0].charAt(0) == 'I'){
                i=Integer.parseInt(tmp[1]);
                maxHeap.offer(i);
                minHeap.offer(i);
            }else if(maxHeap.size()!=0){
                
                if(tmp[1].charAt(0) == '1'){
                    minHeap.remove(maxHeap.poll());
                }else{
                    maxHeap.remove(minHeap.poll());
                }
            }
        }
        if(maxHeap.size()==0){
            answer = new int[]{0,0};
        }else{
            answer = new int[]{maxHeap.poll(),minHeap.poll()};
        }
        
        return answer;
    }
}